department-of-veterans-affairs/caseflow
# Isaiah&MattT/Appeals-4531+Appeals-4532: Add Email Confirmation Field and Validation to VSO Hearings Form + Pre-populate Fields

Resolves [APPEALS-4531](https://vajira.max.gov/browse/APPEALS-4531) and [APPEALS-4532](https://vajira.max.gov/browse/APPEALS-4532)
### Description
This PR combines the branches for APPEALS-4531 and APPEALS-4532 as it became quickly apparent they are closely-tied together.

`4531`: As a VSO User, I need the applicable fields on the convert-to-virtual form (hearings application) to prepopulate with the data from their data source, so the info can be reviewed and/or updated before opting the appellant into a virtual hearing.

`4532`: As a VSO user I need to be prompted to confirm Appellant's email when converting a scheduled hearing to Virtual so that the correct Appellant email is provided.

### Acceptance Criteria
### 4531 (Pre-Population):
- [ ] The following fields on the Convert to Virtual (hearings form) are prepopulated,  with information from the Corporate Database (BGS) when information is available.   
   1. Veteran/Appellant Email 
   2. Veteran/Appellant Time Zone 
   3. POA/Representative Time Zone
- [ ] The following fields will not prepopulate if information is not available in the Corporate Datebase
   1. Veteran/Appellant Email 
   2. Veteran/Appellant Time Zone 
   3. POA/Representative Time Zone
- [ ] The POA/Representative email field is prepopulated for VSO/representative user with the email affiliated with their user's VA account (VA.Gov email)
   1. The POA/Representative email field does not prepopulate for non-VSO user accessing the hearings form
      1. No impact to existing capabilities for this field for non-VSO users
- [ ] The POA/Representative email field is read only for the VSO/representative user
- [ ] The "Optional" label does not appear for VSO users above the read only POA/Representative Email (for these notifications only)
   1. Continues to appear for non-VSO users
- [ ] The POA/Representative Timezone fiel label that reads "Optonal" is updated to "Required"
   1. Appears as required for all users interacting with the form (VSO and non-VSO users)
- [ ] The following fields are required
   1. Veteran/Appellant Email 
   2. Veteran/Appellant Time Zone 
   3. POA/Representative Time Zone

### 4532 (Email Confirmation Field & Validation):
- [ ] A new field is added to the Convert to Virtual Hearing form in the Hearings application
   1. Field Name: Confirm <Appellant/Veteran> Email
      1. <Appellant/Veteran> is dynamic based on appellant type this is how the Hearings form functions today
         1. If appellant is Veteran field will read: Confirm Veteran Email
         2. If appellant is not Veteran field will read: Confirm Appellant Email
   2. Field type: email input field
   3. Validation occurs to ensure proper email format is used
   4. Location: field appears below the Appellant Email (for these notifications only) field
- [ ] The following validations occur for the Confirm Appellant email field:
   1. User is informed valid format is required if attempting to proceed with malformed email address
   2. User is informed if email does not match the Veteran/Appellant Email field
   3. Filed is required for VSO to convert hearing to Virtual
      1. If field inputs do not pass validations then the submission option is disabled
- [ ] The Confirm Appellant Email field is not required for non-VSO users interacting with the form
   1. The Confirm Appellant Email field is hidden for all other non-VSO users interacting with the Convert to Virtual form in the hearings application

### Testing Plan
1. Open the Rails console. You will want to grab both an AMA Hearing and Legacy Hearing for testing purposes:
```ruby
ama_h = Hearing.where(disposition: nil).reject{|h| h.virtual? || h.scheduled_for_past?}.first

legacy_h = LegacyHearing.all.reject{|h| h.virtual? || h.scheduled_for_past?}.first
```
2. Log into Caseflow as a VSO user (ex: `BILLIE_VSO`)

3. First, navigate to the hearing details page for the AMA hearing:
```ruby
ama_h.uuid
```
`/hearings/<ama_h.uuid>/details`

4. The following fields should be pre-populated (assuming the hearing you've ended up has all information present):
   * Veteran/Appellant Email
   * Veteran/Appellant Timezone
   * POA/Representative Email (Read-Only) (Current user's email)
   * POA/Representative Timezone (Current user's timezone)
   
5. Make sure that the submission button in the bottom-right corner of the screen is **disabled** currently. On this branch, this button will remain disabled unless the following criteria are met:
   * Value in primary email address field is present and valid.
   * Confirmation email address matches the primary email address.
   * (Other validations will be brought in on other branches)
   
6. Enter text into the confirmation email field that does not match the primary email field. Click off of the field, and ensure that the proper error message appears.

7. Make the text in the confirmation field match the primary email field. The error should dissipate.

8. Change the primary email field. You should immediately receive an error message on the confirmation email field alerting you that the two fields no longer match.

9. Change the primary email field further if necessary, making the value of the field an invalid email (ex: `lljfl3#!$`). Click off of the field, and you should be met with an error regarding the invalid formatting of the primary email.

10. Place a valid email into the primary email field, and place the same email in the confirmation email field. Ensure that the submission button is now enabled.

11. Fill in the timezone fields, then submit the form.

12. Ensure that the hearing has been converted to virtual:
```ruby
ama_h.reload
ama_h.virtual? (should return => true)
```

13. Repeat these steps for the legacy hearing.
```ruby
legacy_h.vacols_id
```
`/hearings/<legacy_h.vacols_id>/details`

Note that the local database seed for the VACOLS database often is missing appellant information. What is most essential here is that the POA information, which should be that of the current user, is present as to prevent issues with submitting the form as the POA email is required while being read-only to the user.

14. We've made some changes to allow for pre-population of the Queue conversion form. Snag a legacy appeal and navigate to the case details page:
```ruby
la = LegacyAppeal.where(original_hearing_request_type: nil).select { |la| la.hearings.empty? }.first

la.vacols_id
```
`/queue/appeals/<la.vacols_id>`

15. Click on 'Convert to virtual' in the hearings section. Take note of the appellant and representative fields and if they've been pre-populated. It is very possible that the appellant email field is missing a value, but this is more to do with our local database's seed data. What is imperative is that the representative email is populated.

16. Fill out all fields and submit the form. Ensure that the legacy appeal is now how a `changed_hearing_request_type` of virtual:

```ruby
la.reload
la.virtual?
```



### User Facing Changes

 BEFORE|AFTER
 ---|---
![Before Picture](https://user-images.githubusercontent.com/99351305/174890007-38904209-a75e-4200-86dc-533daf79918f.png)|![After Picture](https://user-images.githubusercontent.com/99351305/174890090-a03b7eea-0282-4978-b6ef-a3a1f72edb9e.png)


### Storybook Story
*For Frontend (Presentation) Components*
